docs/multi-currency/multi-currency.md

```diff
@@ -0,0 +1,134 @@
+
+The **(forging2)** rule, together with the **(legitimacy)** rule, ensures that the script `H` (with hash `h`) is run whenever new value of the currency `h` is forged. Because `H` is provided with the `txForge` field of the spending transaction, via the `PendingTx` type, it can block or authorise any amount of value that is created of its currency. There is no need for a registry of currencies because the thing that identifies a currency is (the hash of) its monetary policy.
+
+With this proposal a custom currency is no different from any other smart contract, and currencies don't require a separate cost model.
+
+We can write monetary policy as a state machine that keeps track of the current supply and forges more value when necessary.
+
+By allowing values in the `txForge` field to be negative, we can reduce the supply of a currency. 
+
+## Example (creating a currency)
+
+To illustrate the approach, suppose we have written a validator script `H` that controls the monetary policy of our new currency. To forge a value of `hash(H)`, we need two transactions.
+
+1. `tx1` produces an `o :: TxOut` that is locked with a ByteString  `h = hash(H)`. The `txForge` field of `tx1` is zero.
+2. `tx2` consumes the `TxOut` produced by `tx1`. Its `txForge` field has an entry `{ h -> 1000 }`, so `tx2` forges 1000 units of our currency. For `tx2` to be valid, the `TxIn` that refers to `o` must provide the validator script (ie. the monetary policy) `H` and an appropriate redeemer. The outputs of `tx2` contain (among others) 1000 units of the `h`, which can be spent freely from now on, without referring to the script `H`. `o` is removed from the UTXO set when `tx2` is added to the ledger.
+
+## Example (NFT)
+
+An important example that motivated the search for an alternative to the `CurrencyTx` approach was the idea to implement NFTs as currencies with a supply of 1. If NFTs are used to represent different stakes in a contract, then a single instance of a contract requires multiple NFTs, so it is crucial that currency creation is as lightweight as possible. In our proposal, each NFT adds a single pay-to-script transaction output to the initial transaction of the contract, and  does not otherwise put a burden on core nodes.
+
+Here is an example of a validator script that implements a non-fungible token using our proposal.
+
+```haskell
+data NFT = NFT { nftName :: String, nftBootstrapTxOut :: TxOutput } 
+
+-- the validator script (in reality this would be a quoted TH expression)
+ nftValidator :: NFT -> Data -> Redeemer -> PendingTx -> () 
+ nftValidator (NFT nm txout) _ _ ptx = 
+  let con1 = ptx `spends` txout 
+      con2 = ptx `forges` 1 (ownAdress ptx) 
+  in if con1 && con2 then () else error
```

Comment:
If the ID of the transaction is the hash of its script, then won't this always create the same token? i.e. if you try and run this contract again it's going to try and create a NFT with exactly the same hash, but that already exists.

So we might want some way to "salt" your currencies to avoid clashes due to running the same code several times.

Reply:
`txout` is a transaction output so it has a transaction ID + an index into the transaction's list of outputs. Therefore you can really only run this contract once.

Reply:
To be clear, you can create any number of transactions that produce an output to this contract's address (ie. any number of enabling transactions) but only one of those outputs can be spent, because spending it involves spending `txout` which can only happen once.

Reply:
Yes, I get that. But suppose I run the same code again. I'm using the same script, so it's going to have the same script hash (no?) so it's going to collide with the previous instance.

Reply:
That's "correct", it just might be annoying.

Reply:
If you have currencies with names instead of hashes, then would also get an error if you tried to create the same currency twice. 

In the long term there will be a set of basic contracts (for example a permission system using NFTs, the possibility for a contract to expire / be refunded if nothing happens to it for some time, a state machine, etc.) that you can put together and that hide the annoying bits.

Reply:
Sure, but I think it's a bit more obvious that you need a unique name than that you need to permute your code a bit to get a unique hash!